### PR TITLE
feat(board): add move validation

### PR DESCRIPTION
**Pull request by Carlos-ccaetano**
## Summary

Add a MoveValidator that interprets observed board changes and accepts one legal move for the selected side. This PR is stacked on the board-state work in #85.

History-dependent rules are intentionally outside the current API because the project does not store game history.

## Changes

* validate turn, piece movement, paths, captures, promotion, castling, and en passant;
* reject moves that leave the moving side's king in check;
* accept either two board states or the existing `std::vector<SquareChange>` representation;
* cover valid moves and edge cases with focused tests.

## Testing

* full deterministic acchess suite: 61 tests passed;
* strict MoveValidator build with `-Wall -Wextra -Wpedantic -Werror`: 41 tests passed;
* AddressSanitizer and UndefinedBehaviorSanitizer suite: 61 tests passed;
* root CMake configuration reached an existing SFML/X11 dependency limitation in this headless environment.

## Related

Related to #100

### CODE REVIEW

**Comment by Carlos-ccaetano** [comment]
Fixed in b5c5bdd. Both previous and observed board states now require exactly one king of each color, with regression tests added in 81fb8cf.

File: software/acchess/include/board/move_validator.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * @brief Interprets observed board changes as one legal chess move.
+ *
+ * History-dependent rules are outside this API because the current model does
+ * not store castling rights, prior moves, move counters, or past positions.
+ */
+class MoveValidator {
+public:
+    /**
+     * @return The inferred move, or std::nullopt when the observation is invalid.
```
Comment (oEnzoRibas) [inline software/acchess/include/board/move_validator.hpp:21]: Complementar toda a docstring com "@brief" "@params" e se necessário algum "@note".

File: software/acchess/include/board/move_validator.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    [[nodiscard]] static std::optional<Move> validate(
+        const Board& previous,
+        const Board& observed,
+        PieceColor sideToMove
+    );
+
+    /**
```
Comment (oEnzoRibas) [inline software/acchess/include/board/move_validator.hpp:29]: Complementar toda a docstring com "@brief" "@params" e se necessário algum "@note".

**Review by oEnzoRibas — COMMENTED** [review]
Adicione uma documentação mais completa no move_validator.cpp sobre a implementação mesmo. Enquanto no move_validator.hpp foque apenas no contrato e objetivo das funções.

**Review by oEnzoRibas — CHANGES_REQUESTED** [review]
Adicione uma documentação mais completa no move_validator.cpp sobre a implementação mesmo. Enquanto no move_validator.hpp foque apenas no contrato e objetivo das funções.

**Comment by oEnzoRibas** [comment]
@Carlos-ccaetano Como estão os avanços nas mudanças solicitadas para esse PR?
